server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -1826,6 +1801,45 @@ public void doInTransactionWithoutResult(TransactionStatus status) {
         return vm;
     }
 
+    private void updatePublicIpDnatVmIp(long vmId, long networkId, String oldIp, String newIp) {
+        if (_networkModel.areServicesSupportedInNetwork(networkId, Service.StaticNat)) {
+            List<IPAddressVO> publicIps = _ipAddressDao.listByAssociatedVmId(vmId);
+            for (IPAddressVO publicIp : publicIps) {
+                if (oldIp.equals(publicIp.getVmIp()) && publicIp.getAssociatedWithNetworkId() == networkId) {
+                    publicIp.setVmIp(newIp);
+                    _ipAddressDao.persist(publicIp);
+                }
+            }
+        }
```

Comment:
We could invert the `if` logic and add a return to reduce indentation.

Reply:
@GutoVeronezi
thanks for your suggestion. I will consider it if there are many code lines in the `if` block.
for these 3 methods, they have only 2 lines in `if` block, and I think `&&` is easier to understand than `!(aa) || !(bb)`

Reply:
@weizhouapache sorry, selecting multiple lines may have confused my comment.

 I agree with you when we talk about the second `if` statement, however, I meant invert the first `if` statement, like this:

```suggestion
        if (!_networkModel.areServicesSupportedInNetwork(networkId, Service.StaticNat)) {
            return;
        } 

        List<IPAddressVO> publicIps = _ipAddressDao.listByAssociatedVmId(vmId);
        for (IPAddressVO publicIp : publicIps) {
            if (oldIp.equals(publicIp.getVmIp()) && publicIp.getAssociatedWithNetworkId() == networkId) {
                publicIp.setVmIp(newIp);
                _ipAddressDao.persist(publicIp);
            }
        }
```

Reply:
> @weizhouapache sorry, selecting multiple lines may have confused my comment.
> 
> I agree with you when we talk about the second `if` statement, however, I meant invert the first `if` statement, like this:
> 
> This way we can avoid the `hadouken code` =)

@GutoVeronezi ah ok, my misunderstanding. I will change it. thanks